BUG: Fixed inconsistent offset behavior for series #43784

srotondo opened the pull request:
- [ ] closes #43784 
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

srotondo committed:
BUG: Fixed ignoring of nanoseconds when adding to series #47856

srotondo committed:
Update doc/source/whatsnew/v1.5.0.rst

Co-authored-by: Matthew Roeschke <emailformattr@gmail.com>

srotondo committed:
BUG: Renamed test variables #47856

srotondo committed:
Merge branch 'timedelta-bug' of https://github.com/srotondo/pandas into timedelta-bug

srotondo committed:
BUG: Fixed inconsistent offset behavior for series #43784

srotondo committed:
Merge branch 'main' into dst-bug

srotondo commented:
I'm not exactly sure how these extra commits got into this new PR, but I fixed the file conflicts so that those extra changes are gone.

srotondo committed:
BUG: Fixed code change #43784

srotondo committed:
Merge remote-tracking branch 'upstream/main' into dst-bug

srotondo committed:
Merge branch 'dst-bug' of https://github.com/srotondo/pandas into dst-bug

srotondo commented:
There seems to be a test failing called `test_reindexing_with_float64_NA_log` in `test_reindex.py`, but I've run this test without my code change, and it ends up failing anyways, so I don't think this change is to blame.

srotondo committed:
Merge remote-tracking branch 'upstream/main' into dst-bug

mroeschke commented on doc/source/whatsnew/v1.5.0.rst:
```diff
@@ -992,8 +992,9 @@ Datetimelike
 - Bug in :class:`.DatetimeArray` construction when passed another :class:`.DatetimeArray` and ``freq=None`` incorrectly inferring the freq from the given array (:issue:`47296`)
-- Bug in :func:`to_datetime` where ``OutOfBoundsDatetime`` would be thrown even if ``errors=coerce`` if there were more than 50 rows (:issue:`45319`)
 - Bug when adding a :class:`DateOffset` to a :class:`Series` would not add the ``nanoseconds`` field (:issue:`47856`)
+- Bug where adding a :class:`DateOffset` to a :class:`DatetimeIndex` or :class:`Series` over a Daylight Savings Time boundary would produce an incorrect result (:issue:`43784`)
```
Could you move this to `v1.6.0.rst`?

mroeschke commented on pandas/tests/tseries/offsets/test_dst.py:
```diff
@@ -228,3 +230,27 @@ def test_nontick_offset_with_ambiguous_time_error(original_dt, target_dt, offset
+        [Timestamp("2020-10-26 00:00:00+01:00"), Timestamp("2017-03-13 00:00:00-07:00")]
+    )
+
+    tm.assert_series_equal((startdates + offset1), expected1)
```
Could you assign `(startdates + offset1)` to `result` to make this easier to read?

mroeschke commented on pandas/core/arrays/datetimes.py:
```diff
@@ -714,7 +718,10 @@ def _add_offset(self, offset) -> DatetimeArray:
             if self.tz is not None:
                 # FIXME: tz_localize with non-nano
-                result = result.tz_localize(self.tz)
+                if not offset._use_relativedelta and type(offset) == DateOffset:
```
So this only is needed for `DateOffset` exactly and not its subclasses?

srotondo replied:
Correct.  Almost all of the subclasses of `DateOffset` deal with units larger than an hour, so they would not be affected.  The only ones that could be realistically affected by this would be `BusinessHour` and `CustomBusinessHour`.  However, due to the variable nature of the time differences in those classes (eg. adding 1 hour to the very end of a Friday actually ends up adding over 2 whole days to the start of Monday) I think they should not be affected by this change.

srotondo replied:
I wasn't thinking about user-created subclasses in my original response, rather I was thinking of the predefined subclasses like `YearEnd`.  I have it specifically for `DateOffset` and not its subclasses because this change relies on `_use_relativedelta`, which signifies offsets in units smaller than a day in `DateOffset`.  `DateOffset`'s subclasses have it always set to a default `False`, which I absolutely do not want this change to act on.  Do you think there is a better way to do this?

srotondo committed:
Merge remote-tracking branch 'upstream/main' into dst-bug

srotondo committed:
BUG: Moved release note and changed test #43784

mroeschke commented on doc/source/whatsnew/v1.5.0.rst:
```diff
@@ -994,8 +994,8 @@ Datetimelike
 - Bug in :class:`.DatetimeArray` construction when passed another :class:`.DatetimeArray` and ``freq=None`` incorrectly inferring the freq from the given array (:issue:`47296`)
-- Bug in :func:`to_datetime` where ``OutOfBoundsDatetime`` would be thrown even if ``errors=coerce`` if there were more than 50 rows (:issue:`45319`)
 - Bug when adding a :class:`DateOffset` to a :class:`Series` would not add the ``nanoseconds`` field (:issue:`47856`)
+- Bug in :func:`to_datetime` where ``OutOfBoundsDatetime`` would be thrown even if ``errors=coerce`` if there were more than 50 rows (:issue:`45319`)
```
Could you undo this reordering?

srotondo committed:
BUG: Fixed accidental reordering #43784

srotondo committed:
BUG: Unintentional change #43784

srotondo committed:
BUG: Changed if statement #43784

srotondo commented:
@mroeschke I've changed the check to take into account user-created subclasses by removing the type check in the if statement.  I've also changed the `_use_relativedelta` field in the existing subclasses to match their behavior.

mroeschke commented:
Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.
